### ISSUE TITLE
Suboptimal code generation with nested enums and structs.

### ISSUE THREAD

**Issue by lachlansneff** [rust-lang/rust#55898]
Rust 1.30 with `opt-level=3` compiles

```rust
pub struct Bar {
    l: &'static [u8],
}

pub struct Bar2 {
    l: &'static [u8],
    i: i32,
}

pub struct Bar3 {
    l: &'static [u8],
    i: i32,
    a: i64,
}

pub enum Foo {
    A(Bar),
    B(Bar2),
    C(Bar3),
}

pub fn match_test(foo: Foo) -> usize {
    use Foo::*;
    match foo {
        A(b) => b.l.len(),
        B(b) => b.l.len(),
        C(b) => b.l.len(),
    }
}
```

into

```asm
example::match_test:
        mov     al, byte ptr [rdi]
        cmp     al, 1
        je      .LBB0_2
        cmp     al, 2
.LBB0_2:
        mov     rax, qword ptr [rdi + 16]
        ret
```

as per this godbolt link: https://godbolt.org/z/G67blU.

As far as I can tell, both `cmp` instructions don't need to be there and the whole function should be compiled to just `mov rax, qword ptr [rdi + 16]`.

**Comment by lachlansneff** [comment]
If you change `match_test` to take `Foo` by reference (`&Foo`), the whole function optimizes down to what it should.

**Comment by csmoe** [comment]
Fixed since rustc-1.56.0 cc https://godbolt.org/z/qffcYjj9W
